README.md
# Graph-PHP Graph charts in PHP
graph charts, graph bars, histograms, graph lines, scratter, marks

# V 1.0.1

## CREATE GRAPHS IN PHP:
You can create graphs charts with bars, graphs lines with marks, graphs with background images, histograms, ....

# SCREENSHOTS:
![Screenshot graph lines created in Pure PHP](https://github.com/vivesweb/graph/blob/main/sample%201.png?raw=true)

Nice line graph display with markers

![Screenshot graph bar & lines created in Pure PHP](https://github.com/vivesweb/graph/blob/main/sample2.png?raw=true)

Graph Bars & lines

![Screenshot graph lines created in Pure PHP](https://github.com/vivesweb/graph/blob/main/sample4.png?raw=true)

Nice Math line graph

And now some group of samples:

![Screenshot graphs created in Pure PHP](https://github.com/vivesweb/graph-php/blob/main/multigraph1.jpg?raw=true)
![Screenshot graphs created in Pure PHP](https://github.com/vivesweb/graph-php/blob/main/multigraph2.jpg?raw=true)
![Screenshot graphs created in Pure PHP](https://github.com/vivesweb/graph-php/blob/main/multigraph3.jpg?raw=true)



 # REQUERIMENTS:
 
 Requires packages:
  
  img2img (https://github.com/vivesweb/img2img)
  
  ext-op-ml-php (https://github.com/vivesweb/ext_op_ml)
  
 Requires DeJaVu Fonts: (https://travis-ci.org/dejavu-fonts/dejavu-fonts)
 
 - A minimum (minimum, minimum, minimum requeriments is needed). Tested on:
 		
    - Simple Raspberry pi (B +	512MB	700 MHz ARM11) with Raspbian Lite PHP7.3 (i love this gadgets)  :heart_eyes:
 		
    - VirtualBox Ubuntu Server 20.04.2 LTS (Focal Fossa) with PHP7.4.3

    - Red Hat Enterprise Linux Server release 6.10 (Santiago) PHP Version 7.3.25 (Production Server) 512Mb Memory Limit

    - Red Hat Enterprise Linux release 8.4 (Ootpa). PHP Version 8.0.11 (Production Server) 512Mb Memory Limit

 # SERVER REQUERIMENTS:
 
    - PHP with GD enabled: sudo apt install php-gd
    
    - img2img class: If you need to use IMG_FILTER_SCATTER you need at least PHP >= PHP7.4.0. See official doc: https://www.php.net/manual/en/function.imagefilter.php
    
    - PHP with Imagick for some functions. It is not required for by most methods, but in some, like reading pdf or psd, you will need it. Install: sudo apt install php-imagick
      
    - Requires for Instagram background effects: https://github.com/zaachi/PHP-Instagram-effects


 
# FILES:
 *graph-php.class.php* -> **Main File**.
 
 *example.php* -> **Example File**.
 
 
 # INSTALLATION:
 A lot of easy :smiley:. It is written in PURE PHP. Only need to include the files. Tested on basic PHP installation
 
         require_once( 'graph-php.class.php' );
         

 # NOTES:
 Before draw the Graph, first we need to prepare values inside (as type bar, lines, etc....)
 
# RESUME OF METHODS:

- **CREATE GRAPH OBJECT:**
 
*$graph = new graph( $cfg=null );*

default $cfg:

	[
	'width'     => 6.4, // 6.4 inches
        'height'    => 4.8, // 4.8 inches 
        'dpi'       => 100, // 100 dpis
        'padding'   => .6, // 0.6 inches
        'fontdir'   => __DIR__.'/fonts',
        'fontfamilypath' => 'dejavu-fonts-ttf-2.37/ttf',
        'font'      => 'DejaVuSans.ttf',    
        'fontsize'  => 10.5,   
        'xtickfontsize'  => 10.5,   
        'ytickfontsize'  => 10.5,
        'axes'      => [ 'prop_cycle' => []
            ],
        'lines'     =>
            ['width' => 3],
        'values'     => [],
        'x_values'     => [],
        'y_values'     => [],
        'ylabel'    => '',
        'xlabel'    => '',
        'title'     => '',
        'cycler'    => [ 
            'color' => [ 
                'default' => ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf']
                ],
            'linestyle'  => [ '-', '--', ':', '-.' ]
            ],
        'backgroundstyle'       => 'solid',
        'backgroundcolor'       => '#ffffff',
        'bordertype'            => 'square',
        'paddingleft'           => .79,
        'paddingright'          => .63,
        'paddingtop'            => .58,
        'paddingbottom'         => .515,
        'paddinginsideleft'     => .2,
        'paddinginsideright'    => .2,
        'paddinginsidetop'      => .15,
        'paddinginsidebottom'   => .15,
        'ymarginleftlabel'      => 10,
        'xmarginlabelsticks'    => 16,
        'ymarginlabelsticks'    => 10,
        'xshowlabelticks'       => true,
        'yshowlabelticks'       => true,
        'margintitle'           => 50,
        'x_drawguidelines'      => false,
        'y_drawguidelines'      => false,
		'centerlabels'			=> false,
		'bars'	                => ['percmarginbetwbars' => 20 ],
		'legend'			    => false,
		'legendmarginleft'		=> 7,
        'legendmargintop'		=> 6,
        'legendpaddingleft'		=> 5,
        'legendpaddingtop'		=> 1,
        'legendpaddingright'	=> 5,
        'legendpaddingbottom'	=> 3,
        'legendwidthlines'	    => 28,
        'legendlabelheight'     => 21,
		'width_marker_x'		=> 9,
		'height_marker_x'		=> 9,
		'width_marker_o'		=> 10,
		'height_marker_o'		=> 10,
        'xticks'                => ['rotation' => 0]
		
    ]; // /$default_cfg

Example:

      $graph = new graph();
 
 - **PREPARE BAR GRAPH:**
 
*$graph->bar( $array_values, $arr_values_y_param = null, $cfg = null );*

This method prepare a serie of values in BAR format. Do not echo the graph. To see how to echo graph, see *output_gd_png_base64()* method.

$array_values: Values of Axis X

$arr_values_y_param: Values of Axis Y. If not given, each $array_values will be the Axys Y values and each value will be an index position in Axis X automatically

$cfg: *see CFG_GRAPH_TYPES

Example:

      $graph = new graph();
      $graph->bar( [1, 2, 3, 4] );
      echo '<img src="'.$graph->output_gd_png_base64( ).'" >'; // Echo img raw data in html page
      
With this simple code you will generate the most simplest graph:

![Simple graph bar](https://github.com/vivesweb/graph-php/blob/main/samplesimple.png?raw=true)

Example with x & y values:

      $graph = new graph();
      $graph->bar( [10, 20, 30, 40], [1, 4, 9, 16] );
      $graph->title("With X & Y Values"); // Set the title of the bar. See title() method
      echo '<img src="'.$graph->output_gd_png_base64( ).'" >'; // Echo img raw data in html page
      
With this simple code you will generate the most simplest graph:

![Simple graph bar with X & Y values](https://github.com/vivesweb/graph-php/blob/main/samplexyvalues.png?raw=true)

- **PREPARE LINE GRAPH:**
 
*$graph->plot( $array_values, $arr_values_y_param = null, $cfg = null );*

This method prepare a serie of values in LINES format. Do not echo the graph. To see how to echo graph, see *output_gd_png_base64()* method.

$array_values: Values of Axis X

$arr_values_y_param: Values of Axis Y. If not given, each $array_values will be the Axys Y values and each value will be an index position in Axis X automatically

$cfg: *see CFG_GRAPH_TYPES

Example:

      $graph = new graph();
      $graph->plot( [1, 1.5, 2, 1.8, 3] );
      $graph->title("Simple Plot graph with line"); // Set the title of the bar. See title() method
      echo '<img src="'.$graph->output_gd_png_base64( ).'" >'; // Echo img raw data in html page
      
With this simple code you will generate the most simplest line graph:

![Simple plot line graph](https://github.com/vivesweb/graph-php/blob/main/sampleplotline.png?raw=true)

- **PREPARE MULTILINE GRAPH:**
 
You can prepare multilines calling plot() method as many series as you need.

Example:

      $graph = new graph();
      $graph->plot( [1, 1.5, 2, 1.8, 3] );
      $graph->plot( [2, 2.8, 1.7, 2, 2.3] );
      $graph->title("Multi Line"); // Set the title of the bar. See title() method
      echo '<img src="'.$graph->output_gd_png_base64( ).'" >'; // Echo img raw data in html page
      
With this simple code you will generate Multiline Graph:

![Plot Multiline graph](https://github.com/vivesweb/graph-php/blob/main/samplemultiline.png?raw=true)

- **MERGE LINES & BAR GRAPH:**
 

releases
2021-10-30 v.1.0.1
- Change cfg & gd vars from private to protected for external use in new php-plotlib class
- Change padding automatically when change width or height
- Add new param to cfg to control font size for xticks & yticks
- Add new param to cfg to control margin between X & Y ticklabels & ticks
- Add Boolean draw X & Y label ticks
- Add cfg bordertype [‘square’, ‘halfsquare’] for plot in pairplot
- Add cfg ymarginleftlabel

2021-10-28 v.1.0.0
- First release


example.php
<?php
 /** example.php // example file for graph-php.class.php
  *  
  * 
  * @author Rafael Martin Soto
  * @author {@link https://www.inatica.com/ Inatica}
  * @blog {@link https://rafamartin10.blogspot.com/ Blog Rafael Martin Soto}
  * @since October 2021
  * @version 1.0.0
  * @license GNU General Public License v3.0
  *
  * */

require __DIR__ . '/graph-php.class.php';
?><!doctype html>

<html lang="en">
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Example Graph-PHP</title>
  <meta name="description" content="Example Graph-PHP">
  <meta name="author" content="SitePoint">

  <meta property="og:title" content="Example Graph-PHP">
  <meta property="og:type" content="website">
  <meta property="og:url" content="https://www.inatica.com">
  <meta property="og:description" content="Example Graph-PHP">

</head>
<body>
<?php

//$graph = new graph();
//$max_min = [10, 15];
//print "for ".$max_min[0]."/".$max_min[1]." diff ".($max_min[1]-$max_min[0]).": ";
//echo $graph->compute_offset($max_min[0])."/".$graph->compute_offset($max_min[1]);
//var_dump($graph->compute_offset($max_min) );
  
//unset($graph);



$graph = new graph();
$graph->bar( [1, 2, 3, 4] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();
$graph->title("Here your graph TITLE");
$graph->bar( [1, 2, 3, 4] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->ylabel( 'Here your graph Y LABEL' );
$graph->bar( [1, 2, 3, 4] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->xlabel( 'Here your graph X LABEL' );
$graph->bar( [1, 2, 3, 4] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);




$graph = new graph();

$graph->bar( [10, 20, 30, 40], [1, 4, 9, 16] );
$graph->title("With X & Y Values");
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();

$graph->bar( [1, 2, 3, 4], [1, 4, 9, 16] );
$graph->axes([0, 6, 0, 20]);
$graph->title("Limits Axis X & Y");
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->plot( [1, 1.5, 2, 1.8, 3] );
$graph->title("Simple Plot graph with line");
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->plot( [1, 1.5, 2, 1.8, 3] );
$graph->plot( [2, 2.8, 1.7, 2, 2.3] );
$graph->title("Multi Line");
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$x = $graph->math->linspace( 0, 2, 50 );
$graph->plot( $x, $x, ['label'=>'linear'] );
$graph->plot( $x, $graph->math->pow($x, 2), ['label'=>'quadratic'] );
$graph->plot( $x, $graph->math->pow($x, 3), ['label'=>'cubic'] );
$graph->xlabel('x label');
$graph->ylabel('y label');
$graph->title("Simple Plot. With Legend & Labels X, Y");
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->bar( [1, 2, 3, 4], [10, 9, 10, 8] );
$graph->bar( [1, 2, 3, 4], [8, 6, 9, 7] );
$graph->bar( [1, 2, 3, 4], [6, 5, 7, 5] );
$graph->bar( [1, 2, 3, 4], [3, 3, 4, 2] );
$graph->axes([0, 6, 0, 20]);
$graph->title( 'Multi Bar & fixed Axis Values' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->bar( [1, 2, 3, 4, 5, 6, 7], [1, 4, 9, 16, 17, 18, 17] );
$graph->plot( [1, 2, 3, 4, 5, 6, 7], [10,8, 5, 10,15, 16, 15] );
$graph->title( 'Bar & Line' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->bar( [1, 2, 3, 4, 5, 6, 7], [1, 4, 9, 16, 17, 18, 17] );
$graph->plot( [1, 2, 3, 4, 5, 6, 7], [10,8, 5, 10,15, 16, 15] );
$graph->title( 'Legend' );
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->bar( [1, 2, 3, 4, 5, 6, 7], [1, 4, 9, 16, 17, 18, 17], ['label'=>'Name Legend 1'] );
$graph->plot( [1, 2, 3, 4, 5, 6, 7], [10,8, 5, 10,15, 16, 15], ['label'=>'Name Legend 2'] );
$graph->title( 'Legend with label names' );
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();

$graph->plot( [4, 5, 6, 7], [1, 4, 9, 16], ['marker' => 'x'] );
$graph->plot( [4, 5, 6, 7], [5, 8, 7, 10], ['marker' => 'o'] );
$graph->axes([2, 9, 0, 20]);
$graph->title( 'Lines & Legend with "x" & "o" markers' );
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);

$graph = new graph();

$graph->plot( [4, 5, 6, 7], [1, 4, 9, 16], ['marker' => 'd'] );
$graph->plot( [4, 5, 6, 7], [5, 8, 7, 10], ['marker' => 'd'] );
$graph->axes([2, 9, 0, 20]);
$graph->title( '"d" - Diamond Marker' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->plot( [4, 5, 6, 7], [1, 4, 9, 16], 'o' );
$graph->plot( [4, 5, 6, 7], [5, 8, 7, 10], 'x-' );
$graph->axes([2, 9, 0, 20]);
$graph->title( 'Line with marker & only markers' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);





$graph = new graph();
$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
$graph->title( 'Colors, disctont. line & markers "--", "square", "^"' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->bar( [4.2, 4.4, 4.6, 4.8, 5, 5.2, 5.4, 5.6, 5.8, 6], [1, 4, 9, 16, 1, 4, 9, 16, 5, 4] );
$graph->xlabel( 'entry a' );
$graph->title( 'Histogram of IQ' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->plot( [1, 2, 3, 4], [1, 4, 9, 16] );
$graph->plot( [1, 2, 3, 4], [10, 3, 5, 10] );
$graph->plot( [1, 2, 3, 4], [5, 4, 3, 4] );
$graph->plot( [1, 2, 3, 4], [7, 4, 10, 15] );
$graph->plot( [1, 2, 3, 4], [7, 8, 4, 2] );
$graph->plot( [1, 2, 3, 4], [2, 1, 2, 20] );
$graph->plot( [1, 2, 3, 4], [20, 10, 20, 20] );
$graph->plot( [1, 2, 3, 4], [3, 4, 5, 4] );
$graph->plot( [1, 2, 3, 4], [7, 8, 9, 10] );
$graph->plot( [1, 2, 3, 4], [10, 11, 12, 13] );
$graph->plot( [1, 2, 3, 4], [14, 15, 16, 17] );
$graph->plot( [1, 2, 3, 4], [16, 17, 18, 19] );
$graph->axes([null, null, 0, 20]);
$graph->title( 'A lot of lines' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();
$graph->set_drawguidelines( );

$graph->plot( [1, 2, 3, 4], [1, 4, 9, 16] );
$graph->plot( [1, 2, 3, 4], [10, 3, 5, 10] );
$graph->plot( [1, 2, 3, 4], [5, 4, 3, 4] );
$graph->plot( [1, 2, 3, 4], [7, 4, 10, 15] );
$graph->plot( [1, 2, 3, 4], [7, 8, 4, 2] );
$graph->plot( [1, 2, 3, 4], [2, 1, 2, 20] );
$graph->plot( [1, 2, 3, 4], [20, 10, 20, 20] );
$graph->plot( [1, 2, 3, 4], [3, 4, 5, 4] );
$graph->plot( [1, 2, 3, 4], [7, 8, 9, 10] );
$graph->plot( [1, 2, 3, 4], [10, 11, 12, 13] );
$graph->plot( [1, 2, 3, 4], [14, 15, 16, 17] );
$graph->plot( [1, 2, 3, 4], [16, 17, 18, 19] );
$graph->axes([null, null, 0, 20]);
$graph->title( 'Better with Guidelines' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->plot( [4, 5, 6, 7], [1, 4, 9, 16], ['marker' => __DIR__ . '/author.png',] );
$graph->title( 'Custom image ^_^\'' );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->plot( [4, 5, 6, 7], [1, 4, 9, 16], ['marker' => __DIR__ . '/custom_marker.png', 'label'=>'PHP'] );
$graph->plot( [4, 5, 6, 7], [3, 7, 13, 8], ['marker' => __DIR__ . '/tux.png', 'label'=>'GNU/Linux'] );
$graph->plot( [4, 5, 6, 7], [5, 2, 7, 5], ['marker' => __DIR__ . '/nginx.png', 'label'=>'NGINX'] );
$graph->legendwidthlines( 65 );
$graph->legendlabelheight( 33 );
$graph->title( 'Custom Images with Legend' );
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$graph->plot( [4, 5, 6, 7], [1, 4, 9, 16], ['marker' => __DIR__ . '/custom_marker.png', 'label'=>'PHP'] );
$graph->plot( [4, 5, 6, 7], [3, 7, 13, 8], ['marker' => __DIR__ . '/tux.png', 'label'=>'GNU/Linux'] );
$graph->plot( [4, 5, 6, 7], [5, 2, 7, 5], ['marker' => __DIR__ . '/nginx.png', 'label'=>'NGINX'] );
$graph->set_drawguidelines( );
$graph->legendwidthlines( 65 );
$graph->legendlabelheight( 33 );
$graph->title( 'Better with Guidelines' );
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);





$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$x = $graph->math->linspace( 0, 5, 20 );
$graph->title( 'Original Background' );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);





$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Scaled Bicubic Background' );
$graph->bckgr_img_gd->imagescale( 64, -1, IMG_BICUBIC );
$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Smooth Background' );
$graph->bckgr_img_gd->filter( IMG_FILTER_SMOOTH, -7 );
$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Gaussian Blur Filter Background' );
for($i = 0; $i< 10; $i++ ){
  $graph->bckgr_img_gd->filter( IMG_FILTER_GAUSSIAN_BLUR );
}
$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);







$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Sepia, Brightness & contrast Background' );
$graph->bckgr_img_gd->filter( IMG_FILTER_SEPIA, 4, 80 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, -50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 100 );
$graph->bckgr_img_gd->filter( IMG_FILTER_CONTRAST, 50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 100 );
$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);




$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Gray, Brightness & contrast Background' );
$graph->bckgr_img_gd->filter( IMG_FILTER_GRAYSCALE );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, -50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 100 );
$graph->bckgr_img_gd->filter( IMG_FILTER_CONTRAST, 50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 115 );
$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);




$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'B/W, Brightness & contrast Background' );
$graph->bckgr_img_gd->filter( IMG_FILTER_BLACK_WHITE );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, -50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 100 );
$graph->bckgr_img_gd->filter( IMG_FILTER_CONTRAST, 50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 100 );
$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);




// Interpolation IMG_NEAREST_NEIGHBOUR
$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Scaled & unscaled with Bilinear/G. Blur Background' );

$graph->bckgr_img_gd->imagescale( 64, -1, IMG_BILINEAR_FIXED );
  $graph->bckgr_img_gd->filter( IMG_FILTER_GAUSSIAN_BLUR );
$graph->bckgr_img_gd->imagescale( 400, -1, IMG_BILINEAR_FIXED );

$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



// Interpolation IMG_NEAREST_NEIGHBOUR
$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Scaled & unscaled with Bicubic/G. Blur Background x1' );

$graph->bckgr_img_gd->imagescale( 64, -1, IMG_BICUBIC );
$graph->bckgr_img_gd->filter( IMG_FILTER_GAUSSIAN_BLUR );
$graph->bckgr_img_gd->imagescale( 400, -1, IMG_BICUBIC );

$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



// Interpolation IMG_NEAREST_NEIGHBOUR
$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Scaled & unscaled with Bicubic/G. Blur Background x3' );

$graph->bckgr_img_gd->imagescale( 64, -1, IMG_BICUBIC );
for($i = 0; $i< 3; $i++ ){
  $graph->bckgr_img_gd->filter( IMG_FILTER_GAUSSIAN_BLUR );
}
$graph->bckgr_img_gd->imagescale( 400, -1, IMG_BICUBIC );

$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



// Interpolation IMG_NEAREST_NEIGHBOUR
$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Scaled & unscaled with Bicubic/G. Blur Background x5' );

$graph->bckgr_img_gd->imagescale( 64, -1, IMG_BICUBIC );
for($i = 0; $i< 5; $i++ ){
  $graph->bckgr_img_gd->filter( IMG_FILTER_GAUSSIAN_BLUR );
}
$graph->bckgr_img_gd->imagescale( 400, -1, IMG_BICUBIC );

$x = $graph->math->linspace( 0, 5, 20 );
$graph->plot( [ [$x, $x, 'r--'], [$x, $graph->math->pow($x, 2), 'bs'], [$x, $graph->math->pow($x, 3), 'g^'] ] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$x = $graph->math->linspace( 0, 2, 50 );
$graph->plot( $x, $x, ['label'=>'linear'] );
$graph->plot( $x, $graph->math->pow($x, 2), ['label'=>'quadratic'] );
$graph->plot( $x, $graph->math->pow($x, 3), ['label'=>'cubic'] );
$graph->xlabel('Don\'t use darkness images O_o. Is an example');
$graph->ylabel('y label');
$graph->title("Simple Plot with background & Legend");
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->set_drawguidelines( );
$x = $graph->math->linspace( 0, 2, 50 );
$graph->plot( $x, $x, ['label'=>'linear'] );
$graph->plot( $x, $graph->math->pow($x, 2), ['label'=>'quadratic'] );
$graph->plot( $x, $graph->math->pow($x, 3), ['label'=>'cubic'] );
$graph->xlabel('Horrible, but is an example of what you can do ;D');
$graph->ylabel('y label');
$graph->title("Simple Plot with background, Legend & Guidelines");
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);







$graph = new graph();
$graph->plot( [4, 5, 6, 7], [1, 4, 9, 13], ['marker' => __DIR__ . '/custom_marker.png', 'label'=>'PHP'] );
$graph->plot( [4, 5, 6, 7], [3, 6, 12, 8], ['marker' => __DIR__ . '/tux.png', 'label'=>'GNU/Linux'] );
$graph->plot( [4, 5, 6, 7], [5, 2, 7, 5], ['marker' => __DIR__ . '/nginx.png', 'label'=>'NGINX'] );
$graph->legendwidthlines( 65 );
$graph->legendlabelheight( 33 );
$graph->imread( __DIR__ .'/background_example.jpg');
$graph->title( 'Better use of background' );
$graph->set_drawguidelines( );
$graph->bckgr_img_gd->filter( IMG_FILTER_GRAYSCALE );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, -50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 100 );
$graph->bckgr_img_gd->filter( IMG_FILTER_CONTRAST, 50 );
$graph->bckgr_img_gd->filter( IMG_FILTER_BRIGHTNESS, 115 );
$graph->bckgr_img_gd->flip( );
$graph->legend();
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);







$graph = new graph();
$graph->plot( [40000, 50000, 60000, 70000], [1, 4, 9, 13] );
$graph->title( 'Rotation Xticks' );
$graph->set_drawguidelines( );
$graph->xticks( ['rotation' => 45]);
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$arr_values = [
    [ [1, 1.5, 2, 2, 3, 4], [10, 9.5, 9, 10, 8, 9] ],
    [ [4, 5, 5.7, 6, 7, 8], [8, 6, 7.3, 8, 7, 8] ],
];

$graph->title( 'Scatter' );
$graph->scatter( $arr_values );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);



$graph = new graph();

$arr_values = [
    [ [1, 1.5, 2, 2, 3, 4], [10, 9.5, 9, 10, 8, 9], ['label'=>'Male'] ],
    [ [4, 5, 5.7, 6, 7, 8], [8, 6, 7.3, 8, 7, 8], ['label'=>'Female'] ],
];

$graph->title( 'Scatter & legend' );
$graph->scatter( $arr_values );
$graph->legend( );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->title( 'VALUES FOR HISTOGRAM' );
$graph->bar( [10, 11, 40, 45, 50, 55, 60, 60, 70, 80, 85, 90, 95, 100, 105, 106, 107, 108, 109, 110, 111, 112, 123, 140, 150] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->title( 'DEFAULT HISTOGRAM' );
$graph->hist( [10, 11, 40, 45, 50, 55, 60, 60, 70, 80, 85, 90, 95, 100, 105, 106, 107, 108, 109, 110, 111, 112, 123, 140, 150] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->title( '30 xval HISTOGRAM' );
$graph->hist( [10, 11, 40, 45, 50, 55, 60, 60, 70, 80, 85, 90, 95, 100, 105, 106, 107, 108, 109, 110, 111, 112, 123, 140, 150], ['num_blocks' => 30] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


$graph = new graph();
$graph->title( '7 xval HISTOGRAM' );
$graph->hist( [10, 11, 40, 45, 50, 55, 60, 60, 70, 80, 85, 90, 95, 100, 105, 106, 107, 108, 109, 110, 111, 112, 123, 140, 150], ['num_blocks' => 7] );
?>

  <img src="<?php echo $graph->output_gd_png_base64( );?>">
  <?php
  
unset($graph);


?>
</body>
</html>